Mathlib/Tactic/Linter/RefineLinter.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+#  The "refine" linter
+
+The "refine" linter flags usages of `refine'`.
+
+The tactics `refine` and `refine'` are similar, but they handle meta-variable slightly differently.
+This means that they are not completely interchangeable, nor can one completely replace the other.
+However, `refine` is more readable and (heuristically) tends to be more efficient on average.
+
+This linter is an incentive to discourage uses of `refine'`, without being a ban.
+-/
+
+open Lean Elab
+
+namespace Mathlib.Linter.refine
+
+/-- The refine linter emits a warning on usages of `refine'`. -/
+register_option linter.refine : Bool := {
+  defValue := true
+  descr := "enable the refine linter"
+}
+
+/-- `getRefine' t` returns all usages of `refine'` in the input syntax. -/
+partial
+def getRefine' : Syntax → Array Syntax
+  | stx@(.node _ kind args) =>
+    let rargs := (args.map getRefine').flatten
+    if kind == ``Lean.Parser.Tactic.refine' then rargs.push stx else rargs
```

Comment:
Let me double-check I understood this correctly. (Up to you if you would like to include some of this as comments.)
- If we have a syntax node with arguments, we recurse into the arguments, finding all `refine'`s there.
- If the current node is a `refine'`, we flag that one also.
- The first arm returns `args` (which is empty when no `refine'` is found).
- `t` is something else, it cannot contain any `refine'`

Reply:
I wonder about the last bullet; if this overlooks something. Otherwise, makes sense to me.

Reply:
The other constructors for `Syntax` are `.missing`, `.atom` and `.ident`: the `refine'` tactic nests inside it some of those, but they are all subnodes/leaves of a `.node` with kind `Lean.Parser.Tactic.refine'`.  Conclusion: I do not think that this leaves out anything!